File: haskell/lib/Kan/Ran.hs
module RightKanExtension where

import Control.Applicative (liftA2)
import Data.Functor.Identity (Identity(..))

newtype Ran g h a = Ran { unRan :: forall b. (b -> g a) -> h b -> Ran g h a }

instance Functor (Ran g h) where
  fmap f (Ran ran) = Ran (\k -> fmap f . ran k)

instance Applicative (Ran g h) where
  pure a = Ran (\k -> pure (k (pure a)))
  Ran ranf <*> Ran ranx = Ran (\k -> liftA2 id (ranf (fmap fst k)) (ranx (fmap snd k)))

toRan :: (h a -> t a) -> Ran g h a -> t a
toRan f (Ran ran) = f (ran id)

fromRan :: (Ran g h a -> t a) -> h a -> t a
fromRan s h = s (Ran (\k -> k id h))

adjointToRan :: Adjunction t g -> g a -> Ran t Identity a
adjointToRan Adjunction{..} g = Ran (\k -> k (runIdentity . adjunctUnit g))

ranToAdjoint :: Adjunction t g -> Ran t Identity a -> t a
ranToAdjoint Adjunction{..} (Ran ran) = adjunctRightAdjoint (ran runIdentity)

ranToComposedAdjoint :: Adjunction t g -> Ran t h a -> h (t a)
ranToComposedAdjoint Adjunction{..} (Ran ran) = fmap adjunctRightAdjoint (ran id)

composedAdjointToRan :: Adjunction t g -> h (g a) -> Ran t h a
composedAdjointToRan Adjunction{..} h = Ran (\k -> k adjunctUnit h)

composeRan :: Functor.Composition t g -> Ran t (Ran g h) a -> Ran (Functor.Compose t g) h a
composeRan Functor.Composition{..} (Ran ran) = Ran (\k -> ran (\f -> unCompose . fmap f . compose) k)

decomposeRan :: Functor.Composition t g -> Ran (Functor.Compose t g) h a -> Ran t (h (g a))
decomposeRan Functor.Composition{..} (Ran ran) = Ran (\k -> ran (\f -> k (compose . fmap f . runIdentity)))

gran :: h a -> Ran Identity h a
gran h = Ran (\k -> k runIdentity h)

File: ocaml/lib/kan/ran.ml
type ran g h a = h a -> (a -> g b) -> ran g h b

let fmap f (ran g h a) = ran (f . g) h a  

let (<.*>) (ran1: ran g h a) (ran2: ran g h b) =
  ran (\k -> k (fmap fst k) (fmap snd k)) ((,) <$> ran1 <.> ran2)

let pure a = ran (const a) (pure ())
let (<*>) (ran_f: ran g h (a -> b)) (ran_x: ran g h a) =
  ran (\k -> k (fmap fst k) (fmap snd k)) (liftA2 (flip id) ran_f ran_x)

let to_ran (f: forall a. h a -> t a) (ran: ran g h b) = fmap f (s ran)

let from_ran (s: forall a. ran g h a -> t a) (h_val: h a) = s (gran h_val)

let adjoint_to_ran (adj: Adjunction.t t g) (g_val: g a) = ran (Adjunction.unit adj) (Identity.unit a)

let ran_to_adjoint (adj: Adjunction.t t g) (ran: ran t Identity a) = Adjunction.right_adjoint adj (Identity.run ran)

let ran_to_composed_adjoint (adj: Adjunction.t t g) (ran: ran t h a) =
  fmap (Adjunction.right_adjoint adj) ran

let composed_adjoint_to_ran (adj: Adjunction.t t g) (h_val: h (g a)) =
  ran (Adjunction.unit adj) h_val

let compose_ran (compose: Functor.Composition.t) (ran: ran t (ran g h) a) =
  ran (fmap g . compose . decompose) h a

let decompose_ran (compose: Functor.Composition.t) (ran: ran (compose t g) h a) =
  ran (fmap compose . Identity.unit) (ran h)

let gran (h_val: h a) = ran Identity.id h_val